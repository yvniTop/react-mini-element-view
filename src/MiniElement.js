import React from 'react';
import PropTypes from 'prop-types';

export class MiniElement extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      smallViewMode: false,
      smallViewPos: {
        ...props.initialPosition
      }
    };
    this.originalElm = null;
    this.copyElm = null;
    this.draggable = false;
  }

  componentDidMount() {
    document.addEventListener('scroll', this.onScroll, false);
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.isDraging && !prevState.dragging) {
      document.addEventListener('mousemove', this.onDragMove);
      document.addEventListener('mouseup', this.onDragDone);
    } else if (!this.state.isDraging && prevState.isDraging) {
      document.removeEventListener('mousemove', this.onDragMove);
      document.removeEventListener('mouseup', this.onDragDone);
    }
  }

  onScroll = () => {
    if (!this.originalElm) return false;
    const elmBounding = this.originalElm.getBoundingClientRect();
    const isvisable = this.isInViewPort(elmBounding);

    if (!this.state.smallViewMode) {
      if (!isvisable) {
        this.setState({ smallViewMode: true });
        this.props.onChangeViewMode('miniElement');
      }
    } else {
      const isRealElmVisable = this.isInViewPort(this.copyElm.getBoundingClientRect());

      if (isRealElmVisable) {
        this.setState({ smallViewMode: false });
        this.props.onChangeViewMode('originalElement');
      }
    }

    return isvisable;
  }

  onDragDone = () => {
    document.onmouseup = null;
    document.onmousemove = null;
    const { top, left } = this.originalElm.style;

    this.lastPos = { top, left };
    this.setState({ isDraging: false });
  }

  onStartDrag = (e) => {
    // eslint-disable-next-line
    e = e || window.event;
    e.preventDefault();
    this.lastClientX = e.clientX;
    this.lastClientY = e.clientY;
    this.setState({ isDraging: true });
  }

  onDragMove = (e) => {
    // eslint-disable-next-line
    e = e || window.event;
    const pos1 = this.lastClientX - e.clientX;
    const pos2 = this.lastClientY - e.clientY;
    const maxWidth = document.body.clientWidth;
    let leftPos = this.originalElm.offsetLeft - pos1;
    
    // keep the mini element inside the screen frame
    if (this.originalElm.offsetLeft - pos1 < 0) {
      leftPos = 0;
    }
    if (this.originalElm.offsetLeft - pos1 > maxWidth) {
      leftPos = maxWidth;
    }
    this.lastClientX = e.clientX;
    this.lastClientY = e.clientY;
    this.setState({
      smallViewPos: {
        top: `${(this.originalElm.offsetTop - pos2)}px`,
        left: `${leftPos}px`
      } 
    });
    e.stopPropagation();
    e.preventDefault();
  }

  isInViewPort = (bounding) => (bounding.bottom > 0 && bounding.right > 0);

  render() {
    if (this.state.smallViewMode) {
      const dragEvent = (this.props.draggable) ? {
        onMouseDown: this.onStartDrag
      } : {};

      const { smallViewPos } = this.state;

      return ( 
        <span>  
          <div className="smallWrapper" 
            { ...dragEvent } 
            style={ { 
              position: 'fixed',
               ...this.props.miniElementStyle, 
               ...smallViewPos
            } }
             ref={ (el) => { this.originalElm = el; } } >
            { 
              // eslint-disable-next-line
              this.props.children 
            }
          </div>
          <div 
            style={ {
              height: this.originalElm.offsetHeight, 
              width: this.originalElm.offsetWidth 
              } } 
            ref={ (el) => { this.copyElm = el; } } />
        </span>
      );
    }

    return (   
      <div className="noStyleDiv" style={ { width: 'fit-content', height: 'fit-content' } } ref={ (el) => { this.originalElm = el; } } >
        { this.props.children }
      </div>
    );
  }
}

MiniElement.defaultProps = {
  initialPosition: {
    bottom: '0',
    right: '0'
  },
  miniElementStyle: {
    width: '100px',
    height: '100px'
  },
  onChangeViewMode: () => {},
  draggable: true
};

MiniElement.propTypes = {
  initialPosition: PropTypes.object,
  miniElementStyle: PropTypes.object,
  onChangeViewMode: PropTypes.func,
  draggable: PropTypes.bool
};

export default MiniElement;
