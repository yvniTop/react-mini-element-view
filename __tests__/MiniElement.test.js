import React from 'react';
import jest from 'jest-mock';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';
import { MiniElement } from '../src/MiniElement.js';

Enzyme.configure({ adapter: new Adapter() });

const onChangeViewMode = jest.fn();

const setup = (props = {}) => ({
  ...props,
  onChangeViewMode
});

describe('<MiniElement />', () => {
  const props = setup();
  const wrapper = shallow(
  <MiniElement {...props}>
    <div>Test</div>
  </MiniElement>
  );
  
    test('render', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('initialized `state`', () => {
      expect(wrapper.state()).toEqual({
        smallViewMode: false,
        smallViewPos: {
          bottom: '0',
          right: '0'
        }
      });
    });

    test('initialized props', () => {
      const {
        onChangeViewMode, 
        initialPosition,
        miniElementStyle,
        draggable
        } = wrapper.instance().props;
      expect(draggable).toEqual(true);
      expect(miniElementStyle).toEqual({
        width: '100px',
        height: '100px'
      });
      expect(initialPosition).toEqual({
        bottom: '0',
        right: '0'
      });
      expect(onChangeViewMode).toEqual(onChangeViewMode);
    });
});
