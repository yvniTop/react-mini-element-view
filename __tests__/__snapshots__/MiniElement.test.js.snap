// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MiniElement /> render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MiniElement
    draggable={true}
    initialPosition={
      Object {
        "bottom": "0",
        "right": "0",
      }
    }
    miniElementStyle={
      Object {
        "height": "100px",
        "width": "100px",
      }
    }
    onChangeViewMode={[MockFunction]}
  >
    <div>
      Test
    </div>
  </MiniElement>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        Test
      </div>,
      "className": "noStyleDiv",
      "style": Object {
        "height": "fit-content",
        "width": "fit-content",
      },
    },
    "ref": [Function],
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Test",
      },
      "ref": null,
      "rendered": "Test",
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          Test
        </div>,
        "className": "noStyleDiv",
        "style": Object {
          "height": "fit-content",
          "width": "fit-content",
        },
      },
      "ref": [Function],
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Test",
        },
        "ref": null,
        "rendered": "Test",
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
